File: app/src/main/java/io/github/shadow578/music_dl/db/model/TrackInfo.java
package io.github.shadow578.music_dl.db.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

import java.time.LocalDate;
import java.util.Objects;

import io.github.shadow578.music_dl.KtPorted;
import io.github.shadow578.music_dl.util.storage.StorageKey;

/**
 * information about a track
 */
@Entity(tableName = "tracks",
        indices = {
                @Index("first_added_at")
        })
@KtPorted
public class TrackInfo {

    /**
     * create a new track that is still not downloaded
     *
     * @param id    the youtube id of the track
     * @param title the title of the track
     * @return the track instance
     */
    @NonNull
    public static TrackInfo createNew(@NonNull String id, @NonNull String title) {
        return new TrackInfo(id, System.currentTimeMillis(), title, null, null, null, null, StorageKey.EMPTY, StorageKey.EMPTY, TrackStatus.DownloadPending);
    }

    /**
     * the youtube id of the track
     */
    @NonNull
    @PrimaryKey
    @ColumnInfo(name = "id")
    public final String id;

    /**
     * when this track was first added. millis timestamp, from {@link System#currentTimeMillis()}
     */
    @ColumnInfo(name = "first_added_at")
    public final long firstAddedAt;

    /**
     * the title of the track
     */
    @NonNull
    @ColumnInfo(name = "track_title")
    public String title;

    /**
     * the name of the artist
     */
    @Nullable
    @ColumnInfo(name = "artist_name")
    public String artist;

    /**
     * the day the track was released / uploaded
     */
    @Nullable
    @ColumnInfo(name = "release_date")
    public LocalDate releaseDate;

    /**
     * duration of the track, in seconds
     */
    @Nullable
    @ColumnInfo(name = "duration")
    public Long duration;

    /**
     * the album name, if this track is part of one
     */
    @Nullable
    @ColumnInfo(name = "album_name")
    public String albumName;

    /**
     * the key of the file this track was downloaded to
     */
    @NonNull
    @ColumnInfo(name = "audio_file_key")
    public StorageKey audioFileKey;

    /**
     * the key of the track cover image file
     */
    @NonNull
    @ColumnInfo(name = "cover_file_key")
    public StorageKey coverKey;

    /**
     * is this track fully downloaded?
     */
    @NonNull
    @ColumnInfo(name = "status")
    public TrackStatus status;

    public TrackInfo(@NonNull String id, long firstAddedAt, @NonNull String title, @Nullable String artist, @Nullable LocalDate releaseDate, @Nullable Long duration, @Nullable String albumName, @NonNull StorageKey audioFileKey, @NonNull StorageKey coverKey, @NonNull TrackStatus status) {
        this.id = id;
        this.firstAddedAt = firstAddedAt;
        this.title = title;
        this.artist = artist;
        this.releaseDate = releaseDate;
        this.duration = duration;
        this.albumName = albumName;
        this.audioFileKey = audioFileKey;
        this.coverKey = coverKey;
        this.status = status;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TrackInfo trackInfo = (TrackInfo) o;
        return id.equals(trackInfo.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: app/src/main/kotlin/io/github/shadow578/yodel/db/model/TrackInfo.kt
package io.github.shadow578.yodel.db.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import io.github.shadow578.yodel.util.storage.StorageKey
import java.time.LocalDate

/**
 * information about a track
 */
@Entity(
        tableName = "tracks",
        indices = [
            Index("first_added_at")
        ]
)
data class TrackInfo(
        /**
         * the youtube id of the track
         */
        @ColumnInfo(name = "id")
        @PrimaryKey
        val id: String,

        /**
         * the title of the track
         */
        @ColumnInfo(name = "track_title")
        var title: String,

        /**
         * the name of the artist
         */
        @ColumnInfo(name = "artist_name")
        var artist: String? = null,

        /**
         * the day the track was released / uploaded
         */
        @ColumnInfo(name = "release_date")
        var releaseDate: LocalDate? = null,

        /**
         * duration of the track, in seconds
         */
        @ColumnInfo(name = "duration")
        var duration: Long? = null,

        /**
         * the album name, if this track is part of one
         */
        @ColumnInfo(name = "album_name")
        var albumName: String? = null,

        /**
         * the key of the file this track was downloaded to
         */
        @ColumnInfo(name = "audio_file_key")
        var audioFileKey: StorageKey = StorageKey.EMPTY,

        /**
         * the key of the track cover image file
         */
        @ColumnInfo(name = "cover_file_key")
        var coverKey: StorageKey = StorageKey.EMPTY,

        /**
         * is this track fully downloaded?
         */
        @ColumnInfo(name = "status")
        var status: TrackStatus = TrackStatus.DownloadPending,

        /**
         * when this track was first added. millis timestamp, from [System.currentTimeMillis]
         */
        @ColumnInfo(name = "first_added_at")
        val firstAddedAt: Long = System.currentTimeMillis()
) {
    /**
     * default equals operation. this only checks for the [id].
     * if you want to compare the actual contents of the object, use [equalsContent]
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TrackInfo

        if (id != other.id) return false

        return true
    }

    /**
     * default hashcode implementation. only used [id]
     */
    override fun hashCode(): Int {
        return id.hashCode()
    }

    /**
     * equals operation that takes all fields into account
     */
    fun equalsContent(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TrackInfo

        if (id != other.id) return false
        if (title != other.title) return false
        if (artist != other.artist) return false
        if (releaseDate != other.releaseDate) return false
        if (duration != other.duration) return false
        if (albumName != other.albumName) return false
        if (audioFileKey != other.audioFileKey) return false
        if (coverKey != other.coverKey) return false
        if (status != other.status) return false
        if (firstAddedAt != other.firstAddedAt) return false

        return true
    }


}